File: traceql/clickhouse_transpiler/attr_condition.js
const { getCompareFn, durationToNs, unquote, bitSet } = require('./shared')
const Sql = require('@cloki/clickhouse-sql')
module.exports = class Builder {
  constructor () {
    this.main = null
    this.precondition = null
    this.terms = []
    this.conds = null
    this.aggregatedAttr = ''

    this.sqlConditions = []
    this.isAliased = false
    this.alias = ''
    this.where = []
  }

  /**
   *
   * @param main {BuiltProcessFn}
   * @returns {Builder}
   */
  withMain (main) {
    this.main = main
    return this
  }

  /**
   * @param terms {[]}
   * @returns {Builder}
   */
  withTerms (terms) {
    this.terms = terms
    return this
  }

  /**
   * @param conds
   * @returns {Builder}
   */
  withConditions (conds) {
    this.conds = conds
    return this
  }

  /**
   *
   * @param aggregatedAttr {string}
   * @returns {Builder}
   */
  withAggregatedAttr (aggregatedAttr) {
    this.aggregatedAttr = aggregatedAttr
    return this
  }

  withPrecondition (precondition) {
    this.precondition = precondition
    return this
  }

  /**
   * @returns {ProcessFn}
   */
  build () {
    const self = this
    /** @type {BuiltProcessFn} */
    const res = (ctx) => {
      const sel = self.main(ctx)
      const withPreconditionSel = self.precondition
        ? new Sql.With('precond', self.buildPrecondition(ctx))
        : null
      self.alias = 'bsCond'
      for (const term of self.terms) {
        const sqlTerm = self.getTerm(term)
        self.sqlConditions.push(sqlTerm)
        if (!term.Child('label_name').value.match(/^(\.|span\.|resource\.|name)/)) {
          continue
        }
        self.where.push(sqlTerm)
      }
      const having = self.getCond(self.conds)
      self.aggregator(sel)
      sel.conditions = Sql.And(sel.conditions, Sql.Or(...self.where))
      if (Array.isArray(ctx.randomFilter) && Array.isArray(ctx.cachedTraceIds) && ctx.cachedTraceIds.length > 0) {
        sel.conditions = Sql.And(
          sel.conditions,
          Sql.Or(
            Sql.Eq(new Sql.Raw(`cityHash64(trace_id) % ${ctx.randomFilter[0]}`), Sql.val(ctx.randomFilter[1])),
            new Sql.In('trace_id', 'in', ctx.cachedTraceIds.map(traceId => new Sql.Raw(`unhex('${traceId}')`)))
          ))
      } else if (Array.isArray(ctx.randomFilter)) {
        sel.conditions = Sql.And(
          sel.conditions,
          Sql.Eq(new Sql.Raw(`cityHash64(trace_id) % ${ctx.randomFilter[0]}`), Sql.val(ctx.randomFilter[1])))
      }
      if (withPreconditionSel) {
        sel.with(withPreconditionSel)
        sel.conditions = Sql.And(
          sel.conditions,
          new Sql.In(new Sql.Raw('(trace_id, span_id)'), 'in', new Sql.WithReference(withPreconditionSel)))
      }
      sel.having(having)
      return sel
    }
    return res
  }

  buildPrecondition (ctx) {
    if (!this.precondition) {
      return null
    }
    const sel = this.precondition(ctx)
    sel.select_list = sel.select_list.filter(x => Array.isArray(x) && (x[1] === 'trace_id' || x[1] === 'span_id'))
    sel.order_expressions = []
    return sel
  }

  /**
   * @typedef {{simpleIdx: number, op: string, complex: [Condition]}} Condition
   */
  /**
   * @param c {Token || [any]}
   */
  getCond (c) {
    if (c.name) {
      let left = new Sql.Raw(this.alias)
      if (!this.isAliased) {
        left = groupBitOr(bitSet(this.sqlConditions), this.alias)
      }
      const termIdx = this.terms.findIndex(x => x.value === c.value)
      return Sql.Ne(bitAnd(left, new Sql.Raw((BigInt(1) << BigInt(termIdx)).toString())), Sql.val(0))
    }
    const op = c[0]
    const subs = c.slice(1).map(x => this.getCond(x))
    switch (op) {
      case '&&':
        return Sql.And(...subs)
      case '||':
        return Sql.Or(...subs)
    }
    throw new Error(`unsupported condition operator ${c.op}`)
  }

  /**
   * @param sel {Select}
   */
  aggregator (sel) {
    if (!this.aggregatedAttr) {
      return
    }

    if (this.aggregatedAttr === 'duration') {
      sel.select([new Sql.Raw('toFloat64(any(traces_idx.duration))'), 'agg_val'])
      return
    }

    if (this.aggregatedAttr.match(/^span./)) {
      this.aggregatedAttr = this.aggregatedAttr.substr(5)
    }
    if (this.aggregatedAttr.match(/^resource\./)) {
      this.aggregatedAttr = this.aggregatedAttr.substr(9)
    }
    if (this.aggregatedAttr.match(/^\./)) {
      this.aggregatedAttr = this.aggregatedAttr.substr(1)
    }
    sel.select([sqlAttrValue(this.aggregatedAttr), 'agg_val'])
    this.where.push(Sql.Eq(new Sql.Raw('key'), Sql.val(this.aggregatedAttr)))
  }

  getTerm (term) {
    let key = term.Child('label_name').value
    if (key.match(/^span\./)) {
      key = key.substr(5)
    } else if (key.match(/^resource\./)) {
      key = key.substr(9)
    } else if (key.match(/^\./)) {
      key = key.substr(1)
    } else {
      switch (key) {
        case 'duration':
          return this.getDurationCondition(key, term)
        case 'name':
          key = 'name'
          break
        default:
          throw new Error(`unsupported attribute ${key}`)
      }
    }
    if (term.Child('quoted_str')) {
      return this.getStrCondition(key, term)
    } else if (term.Child('number')) {
      return this.getNumberCondition(key, term)
    }
    throw new Error(`unsupported term statement ${term.value}`)
  }

  getDurationCondition (key, term) {
    const fVal = durationToNs(term.Child('value').value)
    const fn = getCompareFn(term.Child('op').value)
    return fn(new Sql.Raw('traces_idx.duration'), Math.floor(fVal))
  }

  getStrCondition (key, term) {
    const strVal = this.getString(term)
    switch (term.Child('op').value) {
      case '=':
        return Sql.And(
          Sql.Eq(new Sql.Raw('key'), Sql.val(key)),
          Sql.Eq(new Sql.Raw('val'), Sql.val(strVal))
        )
      case '!=':
        return Sql.And(
          Sql.Eq(new Sql.Raw('key'), Sql.val(key)),
          Sql.Ne(new Sql.Raw('val'), Sql.val(strVal))
        )
      case '=~':
        return Sql.And(
          Sql.Eq(new Sql.Raw('key'), Sql.val(key)),
          Sql.Eq(new Sql.Raw(`match(val, ${Sql.quoteVal(strVal)})`), 1)
        )
      case '!~':
        return Sql.And(
          Sql.Eq(new Sql.Raw('key'), Sql.val(key)),
          Sql.Ne(new Sql.Raw(`match(val, ${Sql.quoteVal(strVal)})`), 1)
        )
    }
    throw new Error(`unsupported term statement ${term.value}`)
  }

  getNumberCondition (key, term) {
    const fn = getCompareFn(term.Child('op').value)
    if (!term.Child('value').value.match(/^\d+.?\d*$/)) {
      throw new Error(`invalid value in ${term.value}`)
    }
    const fVal = parseFloat(term.Child('value').value)
    return Sql.And(
      Sql.Eq('key', Sql.val(key)),
      Sql.Eq(new Sql.Raw('isNotNull(toFloat64OrNull(val))'), 1),
      fn(new Sql.Raw('toFloat64OrZero(val)'), fVal)
    )
  }

  getString (term) {
    if (term.Child('quoted_str').value) {
      return unquote(term.Child('quoted_str').value)
    }
    if (term.Child('number').value) {
      return term.Child('number').value
    }
    throw new Error(`unsupported term statement ${term.value}`)
  }
}

/**
 *
 * @param left
 * @param right
 * @returns {SQLObject}
 */
function bitAnd (left, right) {
  const res = new Sql.Raw('')
  res.toString = () => {
    const strLeft = left.toString()
    const strRight = right.toString()
    return `bitAnd(${strLeft}, ${strRight})`
  }
  return res
}

/**
 *
 * @param left
 * @param alias
 * @returns {SQLObject}
 */
function groupBitOr (left, alias) {
  const res = new Sql.Raw('')
  res.toString = () => {
    const strLeft = left.toString()
    if (alias) {
      return `groupBitOr(${strLeft}) as ${alias}`
    }
    return `groupBitOr(${strLeft})`
  }
  return res
}

/**
 *
 * @param attr {string}
 * @returns {SQLObject}
 */
function sqlAttrValue (attr) {
  const res = new Sql.Raw('')
  res.toString = () => {
    const strAttr = Sql.quoteVal(attr)
    return `anyIf(toFloat64OrNull(val), key == ${strAttr})`
  }
  return res
}

/**
 * type sqlAttrValue struct {
 *  attr string
 * }
 *
 * func (s *sqlAttrValue) String(ctx *sql.Ctx, options ...int) (string, error) {
 *  attr, err := sql.NewStringVal(s.attr).String(ctx, options...)
 *  if err != nil {
 *    return "", err
 *  }
 *
 *  return fmt.Sprintf("anyIf(toFloat64OrNull(val), key == %s)", attr), nil
 * }
 */

File: reader/traceql/transpiler/clickhouse_transpiler/attr_condition.go
package clickhouse_transpiler

import (
	"fmt"
	"github.com/metrico/qryn/reader/logql/logql_transpiler_v2/shared"
	traceql_parser "github.com/metrico/qryn/reader/traceql/parser"
	sql "github.com/metrico/qryn/reader/utils/sql_select"
	"strconv"
	"strings"
	"time"
)

type AttrConditionPlanner struct {
	Main           shared.SQLRequestPlanner
	Terms          []*traceql_parser.AttrSelector
	Conds          *condition
	AggregatedAttr string

	sqlConds  []sql.SQLCondition
	isAliased bool
	alias     string
	where     []sql.SQLCondition
}

func (a *AttrConditionPlanner) Process(ctx *shared.PlannerContext) (sql.ISelect, error) {
	main, err := a.Main.Process(ctx)
	if err != nil {
		return nil, err
	}
	a.alias = "bsCond"

	err = a.maybeCreateWhere()
	if err != nil {
		return nil, err
	}

	having, err := a.getCond(a.Conds)
	if err != nil {
		return nil, err
	}

	err = a.aggregator(main)
	if err != nil {
		return nil, err
	}

	res := main.AndWhere(sql.Or(a.where...)).AndHaving(having)

	if ctx.RandomFilter.Max != 0 && len(ctx.CachedTraceIds) > 0 {
		rawCachedTraceIds := make([]sql.SQLObject, len(ctx.CachedTraceIds))
		for i, tid := range ctx.CachedTraceIds {
			rawCachedTraceIds[i] = sql.NewRawObject(fmt.Sprintf("unhex('%s')", tid))
		}
		res.AndWhere(sql.Or(
			sql.Eq(
				sql.NewRawObject(fmt.Sprintf("cityHash64(trace_id) %% %d", ctx.RandomFilter.Max)),
				sql.NewIntVal(int64(ctx.RandomFilter.I)),
			),
			sql.NewIn(sql.NewRawObject("trace_id"), rawCachedTraceIds...),
		))
	} else if ctx.RandomFilter.Max != 0 {
		res.AndWhere(sql.Eq(
			sql.NewRawObject(fmt.Sprintf("cityHash64(trace_id) %% %d", ctx.RandomFilter.Max)),
			sql.NewIntVal(int64(ctx.RandomFilter.I)),
		))
	}

	a.isAliased = false

	return res, nil
}

func (a *AttrConditionPlanner) maybeCreateWhere() error {
	if len(a.sqlConds) > 0 {
		return nil
	}
	for _, t := range a.Terms {
		sqlTerm, err := a.getTerm(t)
		if err != nil {
			return err
		}
		a.sqlConds = append(a.sqlConds, sqlTerm)

		if !strings.HasPrefix(t.Label, "span.") &&
			!strings.HasPrefix(t.Label, "resource.") &&
			!strings.HasPrefix(t.Label, ".") &&
			t.Label != "name" {
			continue
		}
		a.where = append(a.where, sqlTerm)
	}
	return nil
}

func (a *AttrConditionPlanner) aggregator(main sql.ISelect) error {
	if a.AggregatedAttr == "" {
		return nil
	}

	s := main.GetSelect()
	if a.AggregatedAttr == "duration" {
		s = append(s, sql.NewSimpleCol("toFloat64(duration)", "agg_val"))
		main.Select(s...)
		return nil
	}

	if strings.HasPrefix(a.AggregatedAttr, "span.") {
		a.AggregatedAttr = a.AggregatedAttr[5:]
	}
	if strings.HasPrefix(a.AggregatedAttr, "resource.") {
		a.AggregatedAttr = a.AggregatedAttr[9:]
	}
	if strings.HasPrefix(a.AggregatedAttr, ".") {
		a.AggregatedAttr = a.AggregatedAttr[1:]
	}
	s = append(s, sql.NewCol(&sqlAttrValue{a.AggregatedAttr}, "agg_val"))
	main.Select(s...)
	a.where = append(a.where, sql.Eq(sql.NewRawObject("key"), sql.NewStringVal(a.AggregatedAttr)))
	return nil
}

func (a *AttrConditionPlanner) getCond(c *condition) (sql.SQLCondition, error) {
	if c.simpleIdx == -1 {
		subs := make([]sql.SQLCondition, len(c.complex))
		for i, s := range c.complex {
			cond, err := a.getCond(s)
			if err != nil {
				return nil, err
			}
			subs[i] = cond
		}
		switch c.op {
		case "&&":
			return sql.And(subs...), nil
		}
		return sql.Or(subs...), nil
	}
	var left sql.SQLObject
	if !a.isAliased {
		left = &groupBitOr{&bitSet{
			terms: a.sqlConds,
		}, a.alias}
		a.isAliased = true
	} else {
		left = sql.NewRawObject(a.alias)
	}
	return sql.Neq(&bitAnd{left, sql.NewIntVal(int64(1) << c.simpleIdx)}, sql.NewIntVal(0)), nil
}

func (a *AttrConditionPlanner) getTerm(t *traceql_parser.AttrSelector) (sql.SQLCondition, error) {
	key := t.Label
	if strings.HasPrefix(key, "span.") {
		key = key[5:]
	} else if strings.HasPrefix(key, "resource.") {
		key = key[9:]
	} else if strings.HasPrefix(key, ".") {
		key = key[1:]
	} else {
		switch key {
		case "duration":
			return a.getTermDuration(t)
		case "name":
			key = "name"
		default:
			return nil, fmt.Errorf("unsupported attribute %s", key)
		}
	}

	if t.Val.StrVal != nil {
		return a.getTermStr(t, key)
	} else if t.Val.FVal != "" {
		return a.getTermNum(t, key)
	}
	return nil, fmt.Errorf("unsupported statement `%s`", t.String())
}

func (a *AttrConditionPlanner) getTermNum(t *traceql_parser.AttrSelector, key string) (sql.SQLCondition, error) {
	var fn func(left sql.SQLObject, right sql.SQLObject) *sql.LogicalOp
	switch t.Op {
	case "=":
		fn = sql.Eq
	case "!=":
		fn = sql.Neq
	case ">":
		fn = sql.Gt
	case "<":
		fn = sql.Lt
	case ">=":
		fn = sql.Ge
	case "<=":
		fn = sql.Le
	default:
		return nil, &shared.NotSupportedError{Msg: "not supported operator: " + t.Op}
	}

	if t.Val.FVal == "" {
		return nil, fmt.Errorf("%s is not a number value (%s)", t.Val.FVal, t.String())
	}
	fVal, err := strconv.ParseFloat(t.Val.FVal, 64)
	if err != nil {
		return nil, err
	}
	return sql.And(
		sql.Eq(sql.NewRawObject("key"), sql.NewStringVal(key)),
		sql.Eq(sql.NewRawObject("isNotNull(toFloat64OrNull(val))"), sql.NewIntVal(1)),
		fn(sql.NewRawObject("toFloat64OrZero(val)"), sql.NewFloatVal(fVal)),
	), nil
}

func (a *AttrConditionPlanner) getTermStr(t *traceql_parser.AttrSelector, key string) (sql.SQLCondition, error) {
	switch t.Op {
	case "=":
		strVal, err := a.getString(t)
		if err != nil {
			return nil, err
		}
		return sql.And(
			sql.Eq(sql.NewRawObject("key"), sql.NewStringVal(key)),
			sql.Eq(sql.NewRawObject("val"), sql.NewStringVal(strVal)),
		), nil
	case "!=":
		strVal, err := a.getString(t)
		if err != nil {
			return nil, err
		}
		return sql.And(
			sql.Eq(sql.NewRawObject("key"), sql.NewStringVal(key)),
			sql.Neq(sql.NewRawObject("val"), sql.NewStringVal(strVal)),
		), nil
	case "=~":
		strVal, err := a.getString(t)
		if err != nil {
			return nil, err
		}
		return sql.And(
			sql.Eq(sql.NewRawObject("key"), sql.NewStringVal(key)),
			sql.Eq(&matchRe{sql.NewRawObject("val"), strVal}, sql.NewIntVal(1)),
		), nil
	case "!~":
		strVal, err := a.getString(t)
		if err != nil {
			return nil, err
		}
		return sql.And(
			sql.Eq(sql.NewRawObject("key"), sql.NewStringVal(key)),
			sql.Eq(&matchRe{sql.NewRawObject("val"), strVal}, sql.NewIntVal(0)),
		), nil
	}
	return nil, &shared.NotSupportedError{Msg: "not supported operator: " + t.Op}
}

func (a *AttrConditionPlanner) getTermDuration(t *traceql_parser.AttrSelector) (sql.SQLCondition, error) {
	if t.Val.TimeVal == "" {
		return nil, fmt.Errorf("%s is not a time duration value (%s)", t.Val.TimeVal, t.String())
	}
	fVal, err := time.ParseDuration(t.Val.TimeVal)
	if err != nil {
		return nil, err
	}

	fn, err := getComparisonFn(t.Op)
	if err != nil {
		return nil, err
	}

	return fn(sql.NewRawObject("traces_idx.duration"), sql.NewIntVal(fVal.Nanoseconds())), nil
}

func (a *AttrConditionPlanner) getString(t *traceql_parser.AttrSelector) (string, error) {
	var (
		strVal string
		err    error
	)
	if t.Val.StrVal != nil {
		strVal, err = t.Val.StrVal.Unquote()
		if err != nil {
			return "", err
		}
	} else {
		strVal = t.Val.FVal
	}
	return strVal, nil
}

type bitSet struct {
	terms []sql.SQLCondition
}

func (b *bitSet) String(ctx *sql.Ctx, options ...int) (string, error) {
	strTerms := make([]string, len(b.terms))
	for i, term := range b.terms {
		strTerm, err := term.String(ctx, options...)
		if err != nil {
			return "", err
		}
		strTerms[i] = fmt.Sprintf("bitShiftLeft(toUInt64(%s),%d)", strTerm, i)
	}
	res := strings.Join(strTerms, "+")

	return res, nil
}

type bitAnd struct {
	left  sql.SQLObject
	right sql.SQLObject
}

func (b *bitAnd) String(ctx *sql.Ctx, options ...int) (string, error) {

	strLeft, err := b.left.String(ctx, options...)
	if err != nil {
		return "", err
	}
	strRight, err := b.right.String(ctx, options...)
	if err != nil {
		return "", err
	}
	res := fmt.Sprintf("bitAnd(%s,%s)", strLeft, strRight)

	return res, nil
}

type groupBitOr struct {
	left  sql.SQLObject
	alias string
}

func (b *groupBitOr) String(ctx *sql.Ctx, options ...int) (string, error) {

	strLeft, err := b.left.String(ctx, options...)
	if err != nil {
		return "", err
	}
	res := fmt.Sprintf("groupBitOr(%s)", strLeft)
	if b.alias != "" {
		res = fmt.Sprintf("%s as %s", res, b.alias)
	}
	return res, nil
}

type matchRe struct {
	field sql.SQLObject
	re    string
}

func (m matchRe) String(ctx *sql.Ctx, options ...int) (string, error) {
	field, err := m.field.String(ctx, options...)
	if err != nil {
		return "", err
	}
	strRe, err := sql.NewStringVal(m.re).String(ctx, options...)
	if err != nil {
		return "", err
	}
	return fmt.Sprintf("match(%s,%s)", field, strRe), nil
}

type sqlAttrValue struct {
	attr string
}

func (s *sqlAttrValue) String(ctx *sql.Ctx, options ...int) (string, error) {
	attr, err := sql.NewStringVal(s.attr).String(ctx, options...)
	if err != nil {
		return "", err
	}

	return fmt.Sprintf("anyIf(toFloat64OrNull(val), key == %s)", attr), nil
}
